In [15]:
import os
os.chdir("C:\\Users\\lopez\\Desktop\\udea-ai-4-eng-20251-pruebas-saber-pro-colombia") # Ruta donde se encuentra la carpeta con los archivos
os.listdir()

['submission_example.csv', 'test.csv', 'train.csv']

In [29]:
import pandas as pd
df = pd.read_csv("train.csv")
df.info()
df.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 692500 entries, 0 to 692499
Data columns (total 21 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   ID                              692500 non-null  int64  
 1   PERIODO                         692500 non-null  int64  
 2   ESTU_PRGM_ACADEMICO             692500 non-null  object 
 3   ESTU_PRGM_DEPARTAMENTO          692500 non-null  object 
 4   ESTU_VALORMATRICULAUNIVERSIDAD  686213 non-null  object 
 5   ESTU_HORASSEMANATRABAJA         661643 non-null  object 
 6   FAMI_ESTRATOVIVIENDA            660363 non-null  object 
 7   FAMI_TIENEINTERNET              665871 non-null  object 
 8   FAMI_EDUCACIONPADRE             669322 non-null  object 
 9   FAMI_TIENELAVADORA              652727 non-null  object 
 10  FAMI_TIENEAUTOMOVIL             648877 non-null  object 
 11  ESTU_PRIVADO_LIBERTAD           692500 non-null  object 
 12  ESTU_PAGOMATRICU

(692500, 21)

In [17]:
df.head(10)

ID  PERIODO                  ESTU_PRGM_ACADEMICO  \
0  904256    20212                           ENFERMERIA   
1  645256    20212                              DERECHO   
2  308367    20203                MERCADEO Y PUBLICIDAD   
3  470353    20195           ADMINISTRACION DE EMPRESAS   
4  989032    20212                           PSICOLOGIA   
5  659872    20203                 MEDICINA VETERINARIA   
6   47159    20183                  INGENIERIA MECANICA   
7   11829    20183  ADMINISTRACIÓN EN SALUD OCUPACIONAL   
8  257869    20212                INGENIERIA INDUSTRIAL   
9  465511    20183           ADMINISTRACION DE EMPRESAS   

  ESTU_PRGM_DEPARTAMENTO            ESTU_VALORMATRICULAUNIVERSIDAD  \
0                 BOGOTÁ  Entre 5.5 millones y menos de 7 millones   
1              ATLANTICO  Entre 2.5 millones y menos de 4 millones   
2                 BOGOTÁ  Entre 2.5 millones y menos de 4 millones   
3              SANTANDER  Entre 4 millones y menos de 5.5 millones   
4              ANTIOQUIA  Entre 2.5 millones y menos de 4 millones   
5              ANTIOQUIA                         Más de 7 millones   
6                  HUILA  Entre 2.5 millones y menos de 4 millones   
7                 BOGOTÁ    Entre 1 millón y menos de 2.5 millones   
8              ATLANTICO  Entre 5.5 millones y menos de 7 millones   
9              ANTIOQUIA  Entre 2.5 millones y menos de 4 millones   

  ESTU_HORASSEMANATRABAJA FAMI_ESTRATOVIVIENDA FAMI_TIENEINTERNET  \
0       Menos de 10 horas            Estrato 3                 Si   
1                       0            Estrato 3                 No   
2         Más de 30 horas            Estrato 3                 Si   
3                       0            Estrato 4                 Si   
4     Entre 21 y 30 horas            Estrato 3                 Si   
5       Menos de 10 horas            Estrato 5                 Si   
6     Entre 21 y 30 horas            Estrato 2                 Si   
7     Entre 11 y 20 horas            Estrato 2                 Si   
8       Menos de 10 horas            Estrato 1                 Si   
9         Más de 30 horas            Estrato 5                 Si   

                  FAMI_EDUCACIONPADRE FAMI_TIENELAVADORA  ...  \
0    Técnica o tecnológica incompleta                 Si  ...   
1      Técnica o tecnológica completa                 Si  ...   
2  Secundaria (Bachillerato) completa                 Si  ...   
3                             No sabe                 Si  ...   
4                   Primaria completa                 Si  ...   
5      Educación profesional completa                 Si  ...   
6    Educación profesional incompleta                 Si  ...   
7                 Primaria incompleta                 Si  ...   
8  Secundaria (Bachillerato) completa                 Si  ...   
9                           Postgrado                 Si  ...   

  ESTU_PRIVADO_LIBERTAD ESTU_PAGOMATRICULAPROPIO FAMI_TIENECOMPUTADOR  \
0                     N                       No                   Si   
1                     N                       No                   Si   
2                     N                       No                   No   
3                     N                       No                   Si   
4                     N                       No                   Si   
5                     N                       No                   Si   
6                     N                       Si                   Si   
7                     N                       Si                   No   
8                     N                       Si                   Si   
9                     N                       Si                   Si   

  FAMI_TIENEINTERNET.1                   FAMI_EDUCACIONMADRE  \
0                   Si                             Postgrado   
1                   No      Técnica o tecnológica incompleta   
2                   Si    Secundaria (Bachillerato) completa   
3                   Si    Secundaria (Bachillerato

# Rellenar valores null en la columna

Primero verificamos los valores null

In [ ]:
print(df.isnull().sum()) # Verificacion numerica de valores null en toda la tabla

df[df["FAMI_ESTRATOVIVIENDA"].isnull()] # verificacion en tabla de valores null FAMI_ESTRATOVIVIENDA

ID                                    0
PERIODO                               0
ESTU_PRGM_ACADEMICO                   0
ESTU_PRGM_DEPARTAMENTO                0
ESTU_VALORMATRICULAUNIVERSIDAD     6287
ESTU_HORASSEMANATRABAJA           30857
FAMI_ESTRATOVIVIENDA              32137
FAMI_TIENEINTERNET                26629
FAMI_EDUCACIONPADRE               23178
FAMI_TIENELAVADORA                39773
FAMI_TIENEAUTOMOVIL               43623
ESTU_PRIVADO_LIBERTAD                 0
ESTU_PAGOMATRICULAPROPIO           6498
FAMI_TIENECOMPUTADOR              38103
FAMI_TIENEINTERNET.1              26629
FAMI_EDUCACIONMADRE               23664
RENDIMIENTO_GLOBAL                    0
coef_1                                0
coef_2                                0
coef_3                                0
coef_4                                0
dtype: int64


ID  PERIODO                 ESTU_PRGM_ACADEMICO  \
21      284373    20195              INGENIERIA MECATRONICA   
26      925456    20195                             DERECHO   
49      836859    20195   LICENCIATURA EN CIENCIAS SOCIALES   
55      663184    20195                    INGENIERIA CIVIL   
58      629999    20212                      TRABAJO SOCIAL   
...        ...      ...                                 ...   
692444  763147    20183                          OPTOMETRIA   
692448   54207    20183                          PSICOLOGIA   
692454  821212    20195  ADMINISTRACION PUBLICA TERRITORIAL   
692464  592258    20183                 FILOLOGIA E IDIOMAS   
692487  350508    20195                  CONTADURÍA PÚBLICA   

       ESTU_PRGM_DEPARTAMENTO            ESTU_VALORMATRICULAUNIVERSIDAD  \
21                      VALLE  Entre 5.5 millones y menos de 7 millones   
26                    BOLIVAR  Entre 2.5 millones y menos de 4 millones   
49                     BOGOTÁ    Entre 1 millón y menos de 2.5 millones   
55                     TOLIMA  Entre 2.5 millones y menos de 4 millones   
58                      VALLE    Entre 1 millón y menos de 2.5 millones   
...                       ...                                       ...   
692444                 BOGOTÁ  Entre 4 millones y menos de 5.5 millones   
692448                 TOLIMA    Entre 1 millón y menos de 2.5 millones   
692454                 BOGOTÁ                                       NaN   
692464                 BOGOTÁ         Entre 500 mil y menos de 1 millón   
692487                 BOGOTÁ                                       NaN   

       ESTU_HORASSEMANATRABAJA FAMI_ESTRATOVIVIENDA FAMI_TIENEINTERNET  \
21                         NaN                  NaN                NaN   
26                         NaN                  NaN                NaN   
49                         NaN                  NaN                NaN   
55         Entre 21 y 30 horas                  NaN                 Si   
58                         NaN                  NaN                NaN   
...                        ...                  ...                ...   
692444                     NaN                  NaN                NaN   
692448                     NaN                  NaN                NaN   
692454                     NaN                  NaN                NaN   
692464                     NaN                  NaN                NaN   
692487                     NaN                  NaN                NaN   

                     FAMI_EDUCACIONPADRE FAMI_TIENELAVADORA  ...  \
21                                   NaN                NaN  ...   
26                                   NaN                NaN  ...   
49                                   NaN                NaN  ...   
55      Técnica o tecnológica incompleta                 Si  ...   
58                                   NaN                NaN  ...   
...                                  ...                ...  ...   
692444                 Primaria completa                NaN  ...   
692448                               NaN                NaN  ...   
692454                               NaN                NaN  ...   
692464                               NaN                NaN  ...   
692487                               NaN                NaN  ...   

       ESTU_PRIVADO_LIBERTAD ESTU_PAGOMATRICULAPROPIO FAMI_TIENECOMPUTADOR  \
21                         N                       No                  NaN   
26                         N                       No                  NaN   
49                         N                       Si                  NaN   
55                         N                       No                   Si   
58                         N                       No                  NaN   
...                      ...                      ...                  ...   
692444                     N                       No                  NaN   
692448                  

Se observa que tenemos 32137 valores null para la columna FAMI_ESTRATOVIVIENDA


In [23]:
df["FAMI_ESTRATOVIVIENDA"].unique() #Valores en columna

array(['Estrato 3', 'Estrato 4', 'Estrato 5', 'Estrato 2', 'Estrato 1',
       nan, 'Estrato 6', 'Sin Estrato'], dtype=object)

In [24]:
df['FAMI_ESTRATOVIVIENDA'].head(50)

0     Estrato 3
1     Estrato 3
2     Estrato 3
3     Estrato 4
4     Estrato 3
5     Estrato 5
6     Estrato 2
7     Estrato 2
8     Estrato 1
9     Estrato 5
10    Estrato 1
11    Estrato 2
12    Estrato 3
13    Estrato 1
14    Estrato 3
15    Estrato 3
16    Estrato 1
17    Estrato 2
18    Estrato 1
19    Estrato 2
20    Estrato 2
21          NaN
22    Estrato 1
23    Estrato 2
24    Estrato 2
25    Estrato 3
26          NaN
27    Estrato 3
28    Estrato 3
29    Estrato 2
30    Estrato 2
31    Estrato 3
32    Estrato 2
33    Estrato 3
34    Estrato 4
35    Estrato 3
36    Estrato 4
37    Estrato 2
38    Estrato 2
39    Estrato 2
40    Estrato 4
41    Estrato 3
42    Estrato 2
43    Estrato 1
44    Estrato 2
45    Estrato 3
46    Estrato 2
47    Estrato 3
48    Estrato 3
49          NaN
Name: FAMI_ESTRATOVIVIENDA, dtype: object

Rellenamos los valores null de acuerdo a la moda, ya que es un valor categorico

In [25]:
df['FAMI_ESTRATOVIVIENDA'] = df['FAMI_ESTRATOVIVIENDA'].fillna(df['FAMI_ESTRATOVIVIENDA'].mode()[0]) # Rellenamos los valores null con la moda

df['FAMI_ESTRATOVIVIENDA'].head(50)

0     Estrato 3
1     Estrato 3
2     Estrato 3
3     Estrato 4
4     Estrato 3
5     Estrato 5
6     Estrato 2
7     Estrato 2
8     Estrato 1
9     Estrato 5
10    Estrato 1
11    Estrato 2
12    Estrato 3
13    Estrato 1
14    Estrato 3
15    Estrato 3
16    Estrato 1
17    Estrato 2
18    Estrato 1
19    Estrato 2
20    Estrato 2
21    Estrato 2
22    Estrato 1
23    Estrato 2
24    Estrato 2
25    Estrato 3
26    Estrato 2
27    Estrato 3
28    Estrato 3
29    Estrato 2
30    Estrato 2
31    Estrato 3
32    Estrato 2
33    Estrato 3
34    Estrato 4
35    Estrato 3
36    Estrato 4
37    Estrato 2
38    Estrato 2
39    Estrato 2
40    Estrato 4
41    Estrato 3
42    Estrato 2
43    Estrato 1
44    Estrato 2
45    Estrato 3
46    Estrato 2
47    Estrato 3
48    Estrato 3
49    Estrato 2
Name: FAMI_ESTRATOVIVIENDA, dtype: object

Finalmente podemos observar que en los valores 21 y 49 que anteriormente eran valores null, ahora tienen valores de acuerdo a la moda

In [26]:
print(df.isnull().sum())

ID                                    0
PERIODO                               0
ESTU_PRGM_ACADEMICO                   0
ESTU_PRGM_DEPARTAMENTO                0
ESTU_VALORMATRICULAUNIVERSIDAD     6287
ESTU_HORASSEMANATRABAJA           30857
FAMI_ESTRATOVIVIENDA                  0
FAMI_TIENEINTERNET                26629
FAMI_EDUCACIONPADRE               23178
FAMI_TIENELAVADORA                39773
FAMI_TIENEAUTOMOVIL               43623
ESTU_PRIVADO_LIBERTAD                 0
ESTU_PAGOMATRICULAPROPIO           6498
FAMI_TIENECOMPUTADOR              38103
FAMI_TIENEINTERNET.1              26629
FAMI_EDUCACIONMADRE               23664
RENDIMIENTO_GLOBAL                    0
coef_1                                0
coef_2                                0
coef_3                                0
coef_4                                0
FAMI_ESTRATOVIVIENDA_encoded          0
dtype: int64


# Codificación Columna FAMI_ESTRATOVIVIENDA (One-Hot Encoding)

In [27]:
df["FAMI_ESTRATOVIVIENDA"].unique() #Valores en columna

array(['Estrato 3', 'Estrato 4', 'Estrato 5', 'Estrato 2', 'Estrato 1',
       'Estrato 6', 'Sin Estrato'], dtype=object)

In [28]:
df["FAMI_ESTRATOVIVIENDA_encoded"] = pd.Categorical(
    df["FAMI_ESTRATOVIVIENDA"],
    categories=["Sin Estrato", "Estrato 1", "Estrato 2", "Estrato 3", "Estrato 4", "Estrato 5", "Estrato 6"],
    ordered=True,
).codes

df[["FAMI_ESTRATOVIVIENDA", "FAMI_ESTRATOVIVIENDA_encoded"]].drop_duplicates().sort_values("FAMI_ESTRATOVIVIENDA_encoded")


FAMI_ESTRATOVIVIENDA  FAMI_ESTRATOVIVIENDA_encoded
176          Sin Estrato                             0
8              Estrato 1                             1
6              Estrato 2                             2
0              Estrato 3                             3
3              Estrato 4                             4
5              Estrato 5                             5
154            Estrato 6                             6